### DISCUSSION TITLE
Integration with snapserver and 'Bad credentials' issue.

### DISCUSSION THREAD

**Discussion by SteveSpencer** [librespot-org/librespot#1282] [General]
I am running snapcast, with snapserver on a Raspberry Pi 4B, running 64 bit Pi release of bookworm.
I have built librespot, and have the relevant source =  line in the snapserver.conf file.
However, this isn't working, as I am consistently getting an error:

could not initialize spirc: Permission denied { Login failed with reason: Bad credentials }

To eliminate the problem, I then ran librespot outside of SnapCast,

pi@pi64:~/librespot/target/debug $ ./librespot -B pipe -i "0.0.0.0" -p MYPASSWORD -u MYUSERNAME
[2024-05-10T14:38:53Z INFO  librespot] librespot 0.5.0-dev 9959a6c (Built on 2024-05-10, Build ID: kakNgLT4, Profile: debug)
[2024-05-10T14:38:53Z WARN  libmdns] Failed to register IPv6 receiver: Os { code: 97, kind: Uncategorized, message: "Address family not supported by protocol" }
[2024-05-10T14:38:53Z INFO  librespot_playback::mixer::softmixer] Mixing with softvol and volume control: Log(60.0)
[2024-05-10T14:38:53Z INFO  librespot_playback::convert] Converting with ditherer: tpdf
[2024-05-10T14:38:53Z INFO  librespot_playback::audio_backend::pipe] Using StdoutSink (pipe) with format: S16
[2024-05-10T14:38:54Z INFO  librespot_core::session] Connecting to AP "ap-gew1.spotify.com:4070"
[2024-05-10T14:38:54Z ERROR librespot] could not initialize spirc: Permission denied { Login failed with reason: Bad credentials }

The warning is because the machine has ipv6 disabled, so is not unexpected. (Is there a build option I can set to get librespot to NOT use ipv6?)
Am I missing some 'magic' parameter to get this to work?  I have tried a build from the 'master' and 'dev' branches, but both seem to be doing the same?
I'm pretty sure it's nothing to do with snapcast, since I cannot get it to work standalone, so it must be me not calling librespot correctly, or Spotify has changed something.

**Comment by kingosticks** [comment]
The IPv6 error is coming from mDNS which is used for account "discovery". Since you are specifying your username and password on the command line, you are not using discovery and so maybe you can use `--disable-discovery` ?

As for the bad credentials, I can login OK as usual. Do you have a normal Spotify account or something else, e.g. a Facebook one?

**Reply by rwjack** [reply]
How do you pass parameters to the librespot binary through snapserver? My container just does `exec snapserver` and it all works magically, but I'd like to set this up correctly, as you explained

**Reply by kingosticks** [reply]
https://github.com/badaix/snapcast/blob/develop/doc/configuration.md#available-parameters

> params: Optional string appended to the librespot invocation. This allows for arbitrary flags to be passed to librespot, for instance params=--device-type%20avr. The value has to be properly URL-encoded.

**Reply by rwjack** [reply]
respect

**Reply by rwjack** [reply]
Another thing, how would I go about passing multiple parameters to that params flag?

https://github.com/librespot-org/librespot/issues/1308#issuecomment-2301685024

**Reply by kingosticks** [reply]
List them, separated by url-encoded spaces e.g. `params=--device-type%20avr%20--foo`

**Reply by rwjack** [reply]
Ahh, those are URL encoded spaces!

I tried to decode those thinking they were '='. When I didn't get any decoded output I was like wtf is this, especially got confused as to why would people do `params=--device-type=avr=--foo`, but it all makes sense now.

Cheers big fella

**Comment by SteveSpencer** [comment]
Thanks, I will give that a try. The bad credentials issue seems to have resolved itself now, so maybe it was a 'local' issue with something environmental. Either way, after a couple of restarts of snapserver it still seems to be working!
